core/lib/zksync_core/src/fee_model.rs

```diff
@@ -0,0 +1,143 @@
+    provider: Arc<G>,
+    config: MainNodeFeeModelConfig,
+}
+
+impl<G: L1GasPriceProvider + ?Sized> BatchFeeModelInputProvider for MainNodeFeeInputProvider<G> {
+    fn get_fee_model_params(&self) -> MainNodeFeeParams {
+        match self.config {
+            MainNodeFeeModelConfig::V1(config) => MainNodeFeeParams::V1(MainNodeFeeParamsV1 {
+                config,
+                l1_gas_price: self.provider.estimate_effective_gas_price(),
+            }),
+            MainNodeFeeModelConfig::V2(config) => MainNodeFeeParams::V2(MainNodeFeeParamsV2 {
+                config,
+                l1_gas_price: self.provider.estimate_effective_gas_price(),
+                l1_pubdata_price: self.provider.estimate_effective_pubdata_price(),
+            }),
+        }
+    }
+}
+
+impl<G: L1GasPriceProvider + ?Sized> MainNodeFeeInputProvider<G> {
+    pub(crate) fn new(provider: Arc<G>, config: MainNodeFeeModelConfig) -> Self {
+        Self { provider, config }
+    }
+}
+
+/// Calculates the batch fee input based on the main node parameters.
+/// This function uses the `V1` fee model, i.e. where the pubdata price does not include the proving costs.
+pub(crate) fn compute_batch_fee_model_input_v1(
+    params: MainNodeFeeParamsV1,
+    l1_gas_price_scale_factor: f64,
+) -> BatchFeeInput {
+    let l1_gas_price = (params.l1_gas_price as f64 * l1_gas_price_scale_factor) as u64;
+
+    BatchFeeInput::L1Pegged(L1PeggedBatchFeeModelInput {
+        l1_gas_price,
+        fair_l2_gas_price: params.config.minimal_l2_gas_price,
+    })
+}
+
+/// Calculates the batch fee input based on the main node parameters.
+/// This function uses the `V2` fee model, i.e. where the pubdata price does not include the proving costs.
+pub(crate) fn compute_batch_fee_model_input_v2(
+    params: MainNodeFeeParamsV2,
+    l1_gas_price_scale_factor: f64,
+    l1_pubdata_price_scale_factor: f64,
+) -> BatchFeeInput {
+    let MainNodeFeeParamsV2 {
+        config,
+        l1_gas_price,
+        l1_pubdata_price,
+    } = params;
+
+    let MainNodeFeeModelConfigV2 {
+        minimal_l2_gas_price,
+        compute_overhead_percent,
+        pubdata_overhead_percent,
+        batch_overhead_l1_gas,
+        max_gas_per_batch,
+        max_pubdata_per_batch,
+    } = config;
+
+    let l1_gas_price = (l1_gas_price as f64 * l1_gas_price_scale_factor) as u64;
+    let l1_pubdata_price = (l1_pubdata_price as f64 * l1_pubdata_price_scale_factor) as u64;
+
+    // While the final results of the calculations are not expected to have any overflows, the intermediate computations
+    // might, so we use U256 for them.
+    let l1_batch_overhead_wei = U256::from(l1_gas_price) * U256::from(batch_overhead_l1_gas);
+
+    let fair_l2_gas_price = {
+        let l1_batch_overhead_per_gas = l1_batch_overhead_wei / U256::from(max_gas_per_batch);
+        let gas_overhead_wei =
+            (l1_batch_overhead_per_gas.as_u64() as f64 * compute_overhead_percent) as u64;
```

Comment:
add more comments about the logic here - that we are using 'compute_overhead_percent' and 'pubdata_overhead_percent' to 'fairly' distribute the cost of closing the batch between transactions.